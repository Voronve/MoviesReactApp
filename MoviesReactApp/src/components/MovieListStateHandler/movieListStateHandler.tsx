
import './movieListStateHandler.css'
import React, { useState } from 'react';
import { GenreSelect } from '../GenreSelect/genreSelect';
import { MovieList } from '../MovieList/movieList';
import { MovieDetails } from '../MovieDetails/movieDetails';
import { SortControl } from '../SortControl/sortControl';
import { Dialog } from '../Dialog/dialog';


export interface MovieInfo {
    /** Image url*/
    image: string;
    /**Movie title */
    title: string;
    /**Movie release year */
    releaseYear: number;
    /**Movie janres list */
    genres: string[];
    /**Movie rating */
    rating: number;
    /**Movie length */
    duration: string;
    /**Movie description */
    description: string;
}

export enum genres {
    all = "ALL",
    documentary = "DOCUMENTARY",
    comedy = "COMEDY",
    horror = "HORROR",
    crime = "CRIME"
}

interface MovieListStateHandlerProps {
    /** Movie list*/
    moviesList: MovieInfo[],
    /** janres list*/
    janres: string[],
}

/** Movie list functionality section*/
export function MovieListStateHandler({ moviesList, janres }: MovieListStateHandlerProps){
    const [state, setState] = useState({
        moviesList,
        janres,
        selectedGenre: janres[0],
        movieDetailedData: {} as MovieInfo,
        filteredMovies: moviesList });

        const genreClick = (genre: string) => {

            const filteredMovies = genre === genres.all ?
                state.moviesList :
                state.moviesList.filter(movie => movie.genres.includes(genre));

            setState({...state, filteredMovies: filteredMovies, selectedGenre: genre});
        }

        const sortMovieBy = (sortBy: 'default' | 'date' | 'title') => {
            let sortedMoviesList = [];

            switch(sortBy) {
                case 'date':
                    sortedMoviesList = state.filteredMovies.sort( (prev, next) => next.releaseYear - prev.releaseYear );
                    break;
                case 'title':
                    sortedMoviesList = state.filteredMovies.sort( (prev, next) => prev.title.localeCompare(next.title));
                    break;
                default:
                    sortedMoviesList = state.filteredMovies;
            }
            setState({...state, filteredMovies: sortedMoviesList});
        }

        const movieClick = (title: string) => {
            const index = state.moviesList.findIndex(movie => movie.title === title)
            setState({...state, movieDetailedData: state.moviesList[index]});
        }

    return (
        <div className="movieListWrapper">
            <MovieDetails movieData={state.movieDetailedData}/>
            <div className="genreDashboard">
                <GenreSelect genreNameList={state.janres} selectedGenre={state.selectedGenre} onSelect={genreClick}/>
                <SortControl sortBy={sortMovieBy}/>
            </div>
            <MovieList  list={state.filteredMovies} movieClick={movieClick}/>
        </div>
    );
}