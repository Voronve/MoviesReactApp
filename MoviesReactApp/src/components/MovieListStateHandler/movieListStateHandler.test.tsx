import { MovieListStateHandler } from './movieListStateHandler';
import { render } from '@testing-library/react';

describe("Testing GanreSelect component", () => {
    const genreNameList = ['ALL', 'DOCUMENTARY', 'COMEDY', 'HORROR', 'CRIME'];
    const movieListMock = [{
        "image": "Comedy/Barbie.jpg",
        "title": "Barbie",
        "releaseYear": 2023,
        "genres": ["COMEDY"],
        "rating": 8.9,
        "duration": "2h 34min",
        "description": "At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga. Et harum quidem rerum facilis est et expedita distinctio. Nam libero tempore, cum soluta nobis est eligendi optio cumque nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et molestiae non recusandae. Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat."
    },
    {
        "image": "Comedy/GrandBudapest.jpg",
        "title": "GrandBudapest",
        "releaseYear": 2014,
        "genres": ["COMEDY"],
        "rating": 8.9,
        "duration": "2h 34min",
        "description": "At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga. Et harum quidem rerum facilis est et expedita distinctio. Nam libero tempore, cum soluta nobis est eligendi optio cumque nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et molestiae non recusandae. Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat."
    }];

    test("Testing genres rendering ", () => {
        const { asFragment } = render(<MovieListStateHandler moviesList={movieListMock} janres={genreNameList} />);

        expect(asFragment()).toMatchSnapshot();
    });
});