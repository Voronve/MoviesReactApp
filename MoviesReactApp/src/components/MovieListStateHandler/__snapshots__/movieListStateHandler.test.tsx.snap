// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GanreSelect component Testing genres rendering  1`] = `
<DocumentFragment>
  <div
    class="movieListWrapper"
  >
    <div
      class="genreDashboard"
    >
      <div
        class="genresListWrapper"
      >
        <ul
          class="genresList"
        >
          <li
            class="selected"
          >
            ALL
          </li>
          <li
            class=""
          >
            DOCUMENTARY
          </li>
          <li
            class=""
          >
            COMEDY
          </li>
          <li
            class=""
          >
            HORROR
          </li>
          <li
            class=""
          >
            CRIME
          </li>
        </ul>
      </div>
      <div
        class="sortBywrapper"
      >
        <nav>
          <div
            class="sortByText"
          >
            SORT BY
          </div>
          <ul
            class="dropdown-menu"
          >
            <li
              class="active"
            >
              TITLE
            </li>
            <li
              class=""
            >
              RELEASE DATE
            </li>
          </ul>
          <div
            class="arrowDown"
          >
            â–¼
          </div>
        </nav>
      </div>
    </div>
    <div
      class="moviesList"
    >
      <div
        class="movieTile"
      >
        <img
          alt="Barbie"
          src="/media/Comedy/Barbie.jpg"
          width="322px"
        />
        <div
          class="movieInfo"
        >
          <div
            class="flex-container"
          >
            <div
              class="name"
            >
              Barbie
            </div>
            <div
              class="year"
            >
              2023
            </div>
          </div>
          <div
            class="ganres"
          >
            COMEDY
          </div>
        </div>
      </div>
      <div
        class="movieTile"
      >
        <img
          alt="GrandBudapest"
          src="/media/Comedy/GrandBudapest.jpg"
          width="322px"
        />
        <div
          class="movieInfo"
        >
          <div
            class="flex-container"
          >
            <div
              class="name"
            >
              GrandBudapest
            </div>
            <div
              class="year"
            >
              2014
            </div>
          </div>
          <div
            class="ganres"
          >
            COMEDY
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
